### ISSUE TITLE
How to share code between Twinkles on different wikis?

### ISSUE THREAD

**Issue by siddharthvp** [wikimedia-gadgets/twinkle#1226]
In light of #1144, we need to figure out a way through which twinkle installations on different wikis can share code. So the conventional mechanism for "global gadgets" has been to keep the core script on one wiki which the other wikis load, in addition to a local configuration file. This has worked for relatively small gadgets like MoreMenu, ProveIt and HotCat. 

Given Twinkle's sheer size and complexity, would this system work well for us? Site configurations in case of Twinkle can include functions and sometimes even classes. The pressing concern is about breaking changes. In order to improve the core, it may often be necessary to make changes which break existing installations. If the file is loaded dynamically, the existing installations break immediately until the maintainers (if any) fix them. 

A better idea maybe is to not have them load anything dynamically, but make twinkle-core a semantically versioned "library" upon which each wiki can write its configurations. Each twinkle installation would need to have a git repo with twinkle-core as an npm dependency. Updating the core would involve updating the twinkle-core dependency and making any fixes for breaking changes (a migration guide can be provided with every version of the core).

Let me know what your thoughts are. Cc @MusikAnimal @Huji @Xi-Plus @DannyS712 and others who maintain TW on other wikis.

**Comment by siddharthvp** [comment]
On the other hand, with the conventional approach, updates instantly reflect on all installations. Good for the simpler modules like fluff diff and batch* where there isn't much of configuration. And there's no need for their maintainers to use git or npm. 

Wondering if we could have hybrid solutions too ... Loading a fixed oldid version is I guess another way to prevent breaking changes from breaking things immediately.

**Comment by Huji** [comment]
The answer could be that breaking changes should be appropriately communicated (perhaps through the global code) ahead of time.

For instance, a few weeks prior deploying some breaking change onto the global copy, a change is made to it which would log in browser console that such and such function is being deprecated or replaced. Simultanously, through [Tech News](https://meta.wikimedia.org/wiki/Tech/News) or a similar mass delivery system, a message will be sent to the wikis to notify them of the change and explain what they need to do address it. That last part is key: if Twinkle developers cannot be bothered to document how to address breaking changes, then we should not pursue i18n at all.

**Comment by Amorymeltzer** [comment]
I was working on this earlier this fall, before pausing after #1144 preempted it, and basically I was combining the two.  The MoreMenu system is pretty simple and was what I was basing my work for #129 on, but there's no real reason to force every wiki to use GitHub.  What I was crafting was basically having a config box based on sitename, which then loads just what's needed.  That way, the config can be easily hosted here in one place, which also conveniently is a list of places using the main gadget (without which notifications are largely impossible, barring active engagement, of which we've basically none).

There's actually a lot of per-wiki config stuff I've been wanting to do (both part of #129 and not) and setting things up this way means that can all be there as well.  A good answer to that would make what's needed as simple as possible.  Extra functions/classes/whatever can have a dedicated loading area.

As for breaking changes, I guess I'm a bit more sanguine about that.  We've been pretty good about deprecating stuff that breaks backward, and I've done my best at enWiki's WT:TW to post developer-relevant information, but, again, with a dedicated list of who care, outreach is a dead end.  The more likely cases are things driven by MW changes, which aren't always known or prepareable too far ahead of time (although I get hundreds of emails a day just to try!), but that's six-and-one-half at best.

**Comment by siddharthvp** [comment]
> As for breaking changes, I guess I'm a bit more sanguine about that. We've been pretty good about deprecating stuff that breaks backward

As long as we're removing a function entirely or replacing one with another, we're able to handle it. The trouble comes when we want to change the behaviour of an existing function. For this, there isn't really a solution with the dynamic cross-wiki load system. Because of this, we're unable to properly migrate morebits.wiki.page to use promises (#911) or make morebits.wiki.api use format=json by default (#1224). The same problems will occur again with twinkle-core functions. This can be solved by using an npm module (in which we can just make a new major release which need not be back-compatible) or by having clients load permalinked versions only.

**Comment by siddharthvp** [comment]
As there were no later comments, decided to go ahead with the npm module idea. This is most convenient to go along with the module pattern. For gadget-type loads (/w/load.php?modules=xxx), loading permalinked versions isn't even possible.

**Comment by Amorymeltzer** [comment]
I know you strongly disagree, but I continue to think straying further from the wikimedia family is a net negative.

**Comment by siddharthvp** [comment]
Your thinking is based on mere feeling and intuition rather than data. 

If you can be more specific, I can give a more detailed response. At the moment I don't even know what "straying further from the wikimedia family" means. Twinkle has been developed on GitHub since 2011.

**Comment by Amorymeltzer** [comment]
GitHub is convenient for people interested in contributing code; my concern is folks who just want the tool.  The point of all this is not just an exercise in development, it's a means to making editing wikis easier.  There is a huge difference in scope, complication, and dependence on non-WMF accounts and code between "developed on GitHub" and "someone needs to publicly clone a repo, actively manage a node package environment, and stay engaged on an ongoing and rolling basis, including adding access to successors."

**Comment by DannyS712** [comment]
Sorry, just saw this. Until we have global gadgets, the likely process is:
* copy enwiki twinkle, including all of the sub files, from onwiki
* modify accordingly
I think moving anything to NPM just needlessly complicates things. I can't really follow what this proposal is though

**Comment by DannyS712** [comment]
> The answer could be that breaking changes should be appropriately communicated (perhaps through the global code) ahead of time.
> 
> For instance, a few weeks prior deploying some breaking change onto the global copy, a change is made to it which would log in browser console that such and such function is being deprecated or replaced. Simultanously, through [Tech News](https://meta.wikimedia.org/wiki/Tech/News) or a similar mass delivery system, a message will be sent to the wikis to notify them of the change and explain what they need to do address it. That last part is key: if Twinkle developers cannot be bothered to document how to address breaking changes, then we should not pursue i18n at all.

For changes that would only break on enwiki, the current system is probably enough, but if there is a global twinkle that would be broken by BC-incompatible changes, posting at SRM on meta for the attention of global sysops and global interface editors might help with fixing any scripts

**Comment by siddharthvp** [comment]
> There is a huge difference in scope, complication, and dependence on non-WMF accounts and code between "developed on GitHub" and "someone needs to publicly clone a repo, actively manage a node package environment, and stay engaged on an ongoing and rolling basis, including adding access to successors."

@Amorymeltzer Regarding managing an npm package, I think you're missing the fact that it doesn't need any active management at all. This is applicable only for the core - and [a CI workflow is configured](https://github.com/wikimedia-gadgets/twinkle-core/blob/master/.github/workflows/npm-publish.yml) to publish a new NPM version every time a new release is created using the github UI (the deploy token is stored as a GitHub secret). The contributor doesn't even need to have an npm account, let alone needing to have the access to the account under which the package is published. 

re "stay engaged on an ongoing and rolling basis": This is not required. The only reason why they'd need to stay engaged is for the entirely optional step of bumping the core version to take advantage of any enhancements. Using a static copy of a library obviously means that once something works it would keep working (barring exceptional situations like changes in MW itself). On the contrary, active engagement would be required if the core was being live-loaded from another wiki - since it can change any time causing something to break; contributors to the core would also need to be wary of this all the time. Productivity on both sides is hindered.

**Comment by siddharthvp** [comment]
There are pros and cons of both approaches. There is a lot to be said here, but I'll try to be brief. Some arguments are repeated, but I suppose it helps to have all the points in one post. 

Option 1: Core as an npm package
- Problems
  1. Changes to the core aren't immediately reflected across all wikis. While this does prevent breaking changes from breaking things (problem 1 of Option 2), it requires a maintainer to be active for the new enhancements in core to be usable, even if to just run `npm run update; grunt deploy`
  2. In cases when a security bug is found in the core, a fix needs to be deployed on every wiki rather than on just the central wiki. The git repos would need to be updated first to generate the build outputs.
      - Solutions: (i) as long as we follow good code review practices for the core, security bugs needing urgent fixes should never arise in the first place. (ii) Localisers are encouraged to host their git repos as part of wikimedia-gadgets org on GitHub, as this means any one of us can make the fixes on the git side.

- Nice things:
  - This is pretty much the standard approach for any software, even though it's unusual in the MediaWiki ecosystem. You never hear of any software loading live code from another website that could change at any time. Even CDNs deliver a specified versions of the code only.
  - As webpack is used for bundling, it is possible to use code from arbitrary npm packages (such as lodash for instance) in Twinkle. This needs to be very carefully, though – only code whose licensing is compatible with CC-BY-SA/GFDL can be used, and also care should to taken to ensure tree-shaking is taking place and the whole package isn't being included in the build output if only a function or two from the package are used. Not recommended, but possible. 
  - Very optimal user experience as far as loading is concerned. Just one network request loads the entire gadget. Most of the time, it isn't one full request - MW sometimes fetches multiple gadgets in a single request.

Option 2: Core hosted on a central wiki
- Problems:
  1. Breaking changes: It becomes a lot tougher to make changes to twinkle-core as it needs to be ensured that existing live usages don't break. Not being able to do breaking changes is not a good position to be in because that would mean any design mistakes in structuring of classes etc would stick forever.
  2. Unused code loading: Not all wikis would use everything in twinkle-core. Wikis should be loading only what they are actually using. Using an npm package enables a module bundler like Webpack to [tree-shake](https://webpack.js.org/guides/tree-shaking/) (remove) unused functions and classes from the final build output. This is a significant advantage for Option 1 as it means that we add liberally add stuff to the core that may be useful only for a few wikis, without bloating up the code on wikis that aren't using them.
      - Can be mitigated by liberally using a number of files for the core, so that local installations can choose to not load the files whose contents they aren’t using.
  3. Slow loading: A foreign wiki gadget can't be loaded directly via the gadget definition; it needs to be done in the code – an extra web request. If the solution to problem 2 is used, it is a number of extra web requests (for different files).


I am not saying that Option 1 is overwhelmingly better; rather I just wanted to set up things one way or the other – and it's Option 1 that I liked better and whose problems I think are less worrisome. The code as you know is available at https://github.com/wikimedia-gadgets/twinkle-core. Always open to switch to a different architecture if it can convincingly be shown to be better.

**Comment by Amorymeltzer** [comment]
I'm just saying Option 1 understates the problems:

- Every change requires every local maintainer to manually make a change.
    - It also means each and every wiki needs (1) a local intadmin willing to (2) manage a git repo (not part of the job description).
    - As I understand it, it's likely a (3) GitHub repo would be (all but) necessary if there's to be any level input beyond one single individual.  That means managing and public repos, which as far as I know also means agreement to give up certain private information, especially as users fade in and out.
    - Unless development is going to maintain multiple version branches/release cycles/lifetimes, any locally breaking change is *required* in order to get any future updates, no matter how minor, including fixing any *upstream* changes.  There are more of the former, but, it does nothing for the latter.

If any change needs to be manually deployed to every wiki, that's no different than simply copying the text locally.  The fair comparison isn't necessarily to a central wiki code, it's to simply having a good distribution and notification network/system.  Call that 1.5 or 3 or whatever, but it does much the same thing.

**Comment by siddharthvp** [comment]
The deploy target could very well be made a page in userspace from where 
- Twinkle could be used as a user script
- Or an intadmin could copy the code to the MW space

Also on smaller wikis many people who edit scripts a lot tend to be intadmin; not every wiki is enwiki. 

Anything but the simplest gadget out there needs a git repo. If you're opposed to having git repos, that would mean non-trivial configurations need to be written on-wiki which is not an optimal place for making code change suggestions or reviewing changes. Nor is there any possibility of using ES6 or TypeScript or linting tools. Nor is it possible to test changes in any way other than pasting whole files to the console repeatedly . Nor is there any scope for unit testing. This is a system that inherits all the weaknesses of the MediaWiki gadget system, and fails to take advantage of the advances made in the broader JavaScript community. 

Still, that is a workflow that can be supported for wikis that only want to use only fluff/diff/etc which need minimal configs. A dump of the core can posted to some place where it could be used directly. But by no means is this the _only_ workflow that we would support. The whole point of the project is to make non-trivial configurations possible to do in a sane manner. 

Bullet 2 seems to be some FUD about non-existent privacy issues and and as for bullet 3, maintaining multiple version branches (with only one in active dev and the older ones only for bug fixes) is _possible_ and can be considered depending on the situation. I don't really get the whole of what you're saying here and I am becoming increasingly tired of these vague comments (here and #1220) which I've to read 4 times just to be able to guess what you may be saying.

>  The fair comparison isn't necessarily to a central wiki code, it's to simply having a good distribution and notification network/system. Call that 1.5 or 3 or whatever, but it does much the same thing.

You're being vague where it matters the most. What exactly is this option 3? It's easy to criticise anything but difficult to propose alternatives that dont' have the limitations ...
